### ISSUE TITLE
Only write into GRUB_CMDLINE_LINUX_DEFAULT if necessary

### ISSUE THREAD

**Issue by Vogtinator** [OSInside/kiwi#1650]
Some packages and scripts modify `/etc/default/grub` by adding/modifying options in `GRUB_CMDLINE_LINUX_DEFAULT`.

Even when not specifying `kernelcmdline` in `<type/>`, kiwi overwrites the options in the file:

```
[  122s] [ INFO    ]: 10:51:33 | Writing grub2 defaults file
[  122s] [ INFO    ]: 10:51:33 | --> GRUB_CMDLINE_LINUX_DEFAULT:""
[  122s] [ INFO    ]: 10:51:33 | --> GRUB_DISTRIBUTOR:"openSUSE MicroOS"
```

This means that there has to be the full `kernelcmdline` set for each `<type/>`, which also means that for each profile which needs a different cmdline there has to be a separate `<type/>`, which gets messy quickly.

AFAICT there's no workaround available.

The specific use case here is to enable SELinux in some profiles of `openSUSE-MicroOS` images, which requires adding `security=selinux` to the kernel cmdline.

**Comment by schaefi** [comment]
working on this one now

**Comment by Vogtinator** [comment]
Is there a python-kiwi package with this in OBS somewhere so that I can give it a try?

**Comment by schaefi** [comment]
9.23.6 will appear in Virt:App:Staging today

**Comment by Vogtinator** [comment]
Unfortunately I can't get it to build an image properly. Even with `kiwi-image-vmx-requires` added explicitly, it's missing tools like `qemu-img`, `sgdisk` and `kpartx`.

**Comment by schaefi** [comment]
Sorry about this. We have an open doc issue #1686 which we are working on. The suse internal -requires concept has been moved into a generic and upstream accepted concept. This resulted in a set of new packages named "kiwi-systemdeps-XXX".
obs picks them by  a list of provides tags which were ported to the new systemdeps packages. In theory all needed systemdeps
packages should have been selected by their provides tag in obs (kiwi-image:vmx or. kiwi-image:oem in your case).

If this needs fixing in one of your projects the following should be enough to let the worker pull in all it needs:

```
support: kiwi-systemdeps-disk-images
```

You can also look into the prjconf of one of the Virtualization:Appliances:Images:Testing_* projects for further details.

If changes to the systemdeps spec are needed let's discuss them in a new issue and together with @Conan-Kudo 

Hope the info helps to solve your build problem

Thanks

**Comment by Vogtinator** [comment]
The prjconf of `openSUSE:Factory` contains `Prefer: kiwi-image-vmx-requires kiwi-filesystem-requires` which has to be undone and the new ones preferred instead. Let's see.

> You can also look into the prjconf of one of the Virtualization:Appliances:Images:Testing_* projects for further details.

Those all seem to be empty?

**Comment by schaefi** [comment]
> Those all seem to be empty?

There are sub-projects per distro, e.g

https://build.opensuse.org/project/show/Virtualization:Appliances:Images:Testing_x86:suse

**Comment by Vogtinator** [comment]
> The prjconf of `openSUSE:Factory` contains `Prefer: kiwi-image-vmx-requires kiwi-filesystem-requires` which has to be undone and the new ones preferred instead. Let's see.

Yep, works!

> > You can also look into the prjconf of one of the Virtualization:Appliances:Images:Testing_* projects for further details.
> 
> Those all seem to be empty?

Apparently there's another layer of subprojects for some reason. The prjconf in there isn't usable though, as `Support` statements for packages like that would blow up all builds and lead to conflicts.

**Comment by Vogtinator** [comment]
> > The prjconf of `openSUSE:Factory` contains `Prefer: kiwi-image-vmx-requires kiwi-filesystem-requires` which has to be undone and the new ones preferred instead. Let's see.
> 
> Yep, works!

And I can also confirm that the cmdline written by `config.sh` was kept, so the PR works as expected. Thanks!

**Comment by Vogtinator** [comment]
Unfortunately this broke recently, please reopen.

In the log I see:

```
[  131s] [ DEBUG   ]: 14:26:40 | EXEC: [rsync -a --exclude /*.module /usr/src/packages/KIWI-vmx/build/image-root/usr/share/grub2/x86_64-efi/ /usr/src/packages/KIWI-vmx/build/image-root/boot/grub2/x86_64-efi]
[  131s] [ DEBUG   ]: 14:26:40 | EXEC: [blkid /dev/mapper/loop0p3 -s UUID -o value]
[  131s] [ DEBUG   ]: 14:26:40 | EXEC: [blkid /dev/mapper/loop0p3 -s UUID -o value]
[  131s] [ DEBUG   ]: 14:26:40 | EXEC: [grep -q GRUB_USE_LINUXEFI /usr/src/packages/KIWI-vmx/build/image-root/sbin/grub2-mkconfig]
[  131s] [ DEBUG   ]: 14:26:40 | EXEC: [grep -q GRUB_ENABLE_BLSCFG /usr/src/packages/KIWI-vmx/build/image-root/sbin/grub2-mkconfig]
[  131s] [ INFO    ]: 14:26:40 | Writing grub2 defaults file
[  131s] [ INFO    ]: 14:26:40 | --> GRUB_CMDLINE_LINUX_DEFAULT:"rw"
[  131s] [ INFO    ]: 14:26:40 | --> GRUB_DISTRIBUTOR:"openSUSE MicroOS"
[  131s] [ INFO    ]: 14:26:40 | --> GRUB_GFXMODE:auto
```

So `GRUB_CMDLINE_LINUX_DEFAULT` is overwritten with just `rw` instead of keeping it.

**Comment by schaefi** [comment]
strange, don't understand why.... can't remember that there was a change.

Thanks, looking at it

**Comment by schaefi** [comment]
@Vogtinator Hi checked the image description in openSUSE:Factory/openSUSE-MicroOS but only see types that sets a kernelcmdline. Can you point me to the description from where you see the error ?

Thanks

**Comment by Vogtinator** [comment]
Thanks for reopening.

The description I'm building is at https://build.opensuse.org/package/show/home:favogt:microselinux/openSUSE-MicroOS, the `ContainerHost-kvm-and-xen` profile.

Locally I confirmed that the editing of `/etc/default/grub` by `config.sh` works as expected.

**Comment by schaefi** [comment]
@Vogtinator Hi, I found the issue why this happens and created a PR for it.

I'm a little bit concerned why we set the 'rw' option in the default case always. The patch changes this but I don't know why we had this set before. Do you see any negative side effects on the change ?

Thanks

**Comment by schaefi** [comment]
> Thanks! If it's in OBS already I can do a test build as well.

yep it's in Staging. You can take it from here: https://build.opensuse.org/project/show/Virtualization:Appliances:Staging

>It's possible that on some other distributions (or outdated ones), rw is/was necessary, but in those cases that should result in an obvious failure which will be caught early.

yes and those people would now be forced to set "kernelcmdline=rw". Let's hope this is not needed
